### PR TITLE
[Driver][SYCL] Use existing option for device-only and some cleanup

### PR DESCRIPTION
**Pull request by mdtoguchi**
Update -fsycl-device-only usage to be an alias to --offload-device-only. This requires an additional unique check for SYCL enabling due to the expectation that -fsycl-device-only also implies -fsycl.

Update associated tests that use -fsycl-device-only -emit-llvm.  Our default is already bitcode output, fixing this up aligns better with community expectations when using --offload-device-only.

Also do some variable cleanup to better be in sync with community variable names. And use a more common diagnostic for invalid options with -fsycl.

### CODE REVIEW

**Comment by bader** [comment]
> Update associated tests that use -fsycl-device-only -emit-llvm. Our default is already bitcode output, fixing this up aligns better with community expectations when using --offload-device-only.

I find this behavior a bit confusing. My understanding is that by `--offload-device-only` should produce the device code in "executable" format as standard compiler invocation does for the host compilation.
SPIR target has a limitation.  We don't have a native back-end for that target, but we can use SPIR-V translator as SPIR backend compiler to create a SPIR-V format. In that flow, using `--offload-device-only --emit-llvm` is reasonable.
This could be discussed and addressed separately.

**Comment by mdtoguchi** [comment]
> > Update associated tests that use -fsycl-device-only -emit-llvm. Our default is already bitcode output, fixing this up aligns better with community expectations when using --offload-device-only.
> 
> I find this behavior a bit confusing. My understanding is that by `--offload-device-only` should produce the device code in "executable" format as standard compiler invocation does for the host compilation. SPIR target has a limitation. We don't have a native back-end for that target, but we can use SPIR-V translator as SPIR backend compiler to create a SPIR-V format. In that flow, using `--offload-device-only --emit-llvm` is reasonable. This could be discussed and addressed separately.

To support the ability to emit SPIR-V using the translator, we have `-fsycl-device-only -fsycl-device-obj=spirv`.  Our default is `-fsycl-device-obj=llvmir`

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -5220,8 +5220,8 @@ static void ProcessVSRuntimeLibrary(const ToolChain &TC, const ArgList &Args,
       (RTOptionID == options::OPT__SLASH_MT ||
        RTOptionID == options::OPT__SLASH_MTd))
     // Use of /MT or /MTd is not supported for SYCL.
-    TC.getDriver().Diag(diag::err_drv_unsupported_opt_sycl)
-        << SetArg->getOption().getName();
+    TC.getDriver().Diag(clang::diag::err_drv_argument_not_allowed_with)
```
[the remark this answers is not in the document]
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:5223]: Yup, you're right.  I'll clean it up.

**Comment by bader** [comment]
> To support the ability to emit SPIR-V using the translator, we have `-fsycl-device-only -fsycl-device-obj=spirv`. Our default is `-fsycl-device-obj=llvmir`

Right. The primary reason for that is to support Relocatable Device Code (`-fgpu-rdc`) by default. We keep device code in LLVM format to be able to link translation units with device code.

My point is that community compiler uses `no-gpu-rdc` by default (see https://clang.llvm.org/docs/HIPSupport.html#device-code-compilation). Potentially SPIR-V is also "linkable" file format, but we don't have a production quality SPIR-V linker tool.

We can also consider switching to `no-gpu-rdc` for SYCL mode by the default as well. This should be a separate discussion, so, please, do not consider my comments as blocking.

**Review by sergey-semenov — APPROVED** [review]

**Review by sarnex — APPROVED** [review]

**Review by hchilama — APPROVED** [review]

**Comment by mdtoguchi** [comment]
@intel/llvm-gatekeepers, this should be ready for merge, thanks!

**Pull request merged**